### PR TITLE
client, fix Jackson de-serialize error when ctor has mismatched client type and wire type

### PR DESCRIPTION
**Pull request by weidongxu-microsoft**
fix https://github.com/Azure/autorest.java/issues/2170

this commit is a sample of the changes on generated code https://github.com/Azure/autorest.java/pull/2202/commits/f4612132421c618c6a25fc3a51b6fc0a5cbefe52

### CODE REVIEW

File: javagen/src/main/java/com/azure/autorest/template/ModelTemplate.java [hunk 1]
```diff
@@ -762,6 +792,52 @@ private void addModelConstructor(ClientModel model, JavaVisibility constructorVi
...
+
+                addModelConstructorParameterAsWireType(property, constructorPropertiesAsWireType);
+
+                addParameterInvokePublicConstructor.accept(property);
+            }
+
+            classBlock.privateConstructor(model.getName() + "(" + constructorPropertiesAsWireType + ")", constructor -> {
+                constructor.line("this(" + constructorPropertiesInvokePublicConstructor + ");");
+            });
+        }
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/template/ModelTemplate.java:840]: Main logic here, to generate another private ctor with parameters as wire types.

File: javagen/src/main/java/com/azure/autorest/template/ModelTemplate.java [hunk 2]
```diff
@@ -684,27 +685,55 @@ private void addModelConstructor(ClientModel model, JavaVisibility constructorVi
...
+                        return true;
+                    }
+                }
+            };
+            final boolean constructorParametersContainsMismatchWireType =
+                requiredProperties.stream().anyMatch(isWireTypeMismatch) || requiredParentProperties.stream().anyMatch(isWireTypeMismatch);
+
+            if (constructorParametersContainsMismatchWireType && !settings.isStreamStyleSerialization()) {
+                generatePrivateConstructorForJackson = true;
+            }
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/template/ModelTemplate.java:726]: Only do this, when there is mismatch, AND it is not stream-stype.

File: typespec-tests/src/main/java/com/encode/datetime/models/Rfc7231DatetimeProperty.java [hunk 3]
```diff
@@ -27,11 +27,16 @@ public final class Rfc7231DatetimeProperty {
...
+    public Rfc7231DatetimeProperty(OffsetDateTime value) {
         this.value = new DateTimeRfc1123(value);
     }
 
+    @Generated
+    @JsonCreator
+    private Rfc7231DatetimeProperty(@JsonProperty(value = "value") DateTimeRfc1123 value) {
+        this(value.getDateTime());
+    }
+
```
Comment (weidongxu-microsoft) [inline typespec-tests/src/main/java/com/encode/datetime/models/Rfc7231DatetimeProperty.java:39]: change in generated code like this

**Comment by weidongxu-microsoft** [comment]
Latest cadl-ranch actually had a case on generic type (List) that we cannot handle even after this PR.

https://github.com/Azure/autorest.java/blob/9bfe7aa16a9129c57f7faab18230de954a321ff4/typespec-tests/src/main/java/com/encode/bytes/models/Base64UrlArrayBytesProperty.java#L29-L33
which would requires a ctor with `List<Base64Url>`.

In this case, the class does not have superclass or subclass. We can generate a empty private class for Jackson.
But this will not work in more complex case.

File: javagen/src/main/java/com/azure/autorest/template/ModelTemplate.java [hunk 4]
```diff
@@ -684,27 +685,55 @@ private void addModelConstructor(ClientModel model, JavaVisibility constructorVi
...
         // If there are no required constructor properties this will simply be zero and result in a no-args constructor
         // being generated.
         StringBuilder constructorProperties =
-            new StringBuilder(128 * (requiredProperties.size() + requiredParentProperties.size()));
+            new StringBuilder(constructorPropertiesStringBuilderCapacity);
 
         StringBuilder superProperties = new StringBuilder(64 * requiredParentProperties.size());
 
         if (settings.isRequiredFieldsAsConstructorArgs()) {
+            final Predicate<ClientModelProperty> isWireTypeMismatch = p -> {
```
Comment (alzimmermsft) [inline javagen/src/main/java/com/azure/autorest/template/ModelTemplate.java:707]: Could we turn this into a utility method? I imagine we have other places checking for `wireType != clientType`. Might want to design this utility method with an additional boolean parameter on whether to consider `GenericType`s as equalivalent.

**Review by alzimmermsft — APPROVED** [review]

File: javagen/src/main/java/com/azure/autorest/template/ModelTemplate.java (hunk 4 above)
Comment (weidongxu-microsoft) [reply javagen/src/main/java/com/azure/autorest/template/ModelTemplate.java:707]: sure

**Pull request merged**